Update doctrine/dbal to v3

craig410 opened the pull request:
* [x] Changelog

craig410 committed:
Update doctrine/dbal to v3

craig410 committed:
:bomb: use doctrine native PDO connection, not wrapped

It looks like the only option is to use the ->getNativeConnection if
we need something that passes `$conn instanceof \PDO`. This will mean
that exception handling changes.

NB that this means we need at least doctrine/dbal@3.3 as
getNativeConnection was not present before that release.

acoulton committed:
Implement a fake mysql driver for use in test envs

acoulton commented:
@craig410 I think this may need more but need to play with it in a project to get a better understanding of how we actually use it / what the driver stack looks like at runtime with a real DB (they have so many classes proxying to others it's quite hard to follow).

I need also to work out how big a deal / the implications of no longer having access to the wrapped connection as opposed to the native one, so we should keep using wrapped for now even tho it's deprecated.

So don't release yet but should be able to pull this branch in as a dev dependency for now.

acoulton committed:
Have NullPDO return a value for getAttribute(PDO::ATTR_SERVER_VERSION)

Doctrine needs to read a server version to create the database
platform driver, even if it is not actually checking database type
definitions. For example, in order to check class metadata.

The actual value is somewhat unimportant, so long as it is a valid
supported mysql version.

The outer connection class has the option to return a `null` if
the wrapped connection does not support getting a version, but
because we are using Doctrine's PDO wrapper that class indicates
that it does provide a version and can only return a string.

acoulton committed:
Update CHANGELOG and README
